server/app/services/geo/esri/EsriClient.java

```diff
@@ -0,0 +1,204 @@
+package services.geo.esri;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import com.typesafe.config.Config;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import javax.inject.Inject;
+import org.apache.commons.lang3.mutable.MutableInt;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import play.libs.Json;
+import play.libs.ws.WSBodyReadables;
+import play.libs.ws.WSBodyWritables;
+import play.libs.ws.WSClient;
+import play.libs.ws.WSRequest;
+import play.libs.ws.WSResponse;
+import services.Address;
+import services.AddressField;
+import services.geo.AddressLocation;
+import services.geo.AddressSuggestion;
+import services.geo.AddressSuggestionGroup;
+
+/**
+ * Provides methods for handling reqeusts to external Esri geo and map layer services for getting
+ * address suggestions and service area validation for a given address.
+ *
+ * <p>@see <a
+ * href="https://developers.arcgis.com/rest/geocode/api-reference/geocoding-find-address-candidates.htm">Find
+ * Address Candidates</a>
+ *
+ * @see <a
+ *     href="https://developers.arcgis.com/rest/services-reference/enterprise/query-feature-service-layer-.htm">Query
+ *     (Map Service/Layer)</a>
+ */
+public class EsriClient implements WSBodyReadables, WSBodyWritables {
+  private final WSClient ws;
+
+  public static final String ESRI_CONTENT_TYPE = "application/json";
+  // Specify output fields to return in the geocoding response with the outFields parameter
+  public static final String ESRI_FIND_ADDRESS_CANDIDATES_OUT_FIELDS =
+      "Address, SubAddr, City, Region, Postal";
+  // The service supports responses in JSON or PJSON format. You can specify the response format
+  // using the f parameter. This is a required parameter
+  public static final String ESRI_FIND_ADDRESS_CANDIDATES_FORMAT = "json";
+  public Optional<String> ESRI_FIND_ADDRESS_CANDIDATES_URL;
+  public int ESRI_FIND_ADDRESS_CANDIDATES_TRIES;
+
+  private final Logger logger = LoggerFactory.getLogger(this.getClass());
+
+  @Inject
+  public EsriClient(Config configuration, WSClient ws) {
+    this.ws = checkNotNull(ws);
+    this.ESRI_FIND_ADDRESS_CANDIDATES_URL =
+        Optional.ofNullable(configuration.getString("esri_find_address_candidates_url"));
+    this.ESRI_FIND_ADDRESS_CANDIDATES_TRIES =
+        Optional.ofNullable(configuration.getInt("esri_find_address_candidates_tries")).orElse(3);
+  }
```

Comment:
issue (blocking): this code will throw an exception rather than return `null` if there's no value present for the given keys

https://github.com/lightbend/config#how-to-handle-defaults

Reply:
Updated to use Config.hasPath()